### DISCUSSION TITLE
Osu Lazer won't launch due to SteamVR.

### DISCUSSION THREAD

**Discussion by M4rkFlor** [ppy/osu#20481] [Q&A] [answered]
I open lazer and nothing happens. normal osu! still launches normally. SteamVR seems to be preventing Osu! Lazer from launching.

Logs
[legacy-ipc.log](https://github.com/ppy/osu/files/9643571/legacy-ipc.log)
[performance.log](https://github.com/ppy/osu/files/9643572/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9643573/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9643574/updater.log)
[database.log](https://github.com/ppy/osu/files/9643575/database.log)

**Comment by peppy** [comment]
It looks like osu! is having trouble determining your monitor layout. Could you try disabling all but one monitor and see if the game starts as expected?

**Reply by M4rkFlor** [reply]
I unplugged all monitors and and vr headset then turn off/on pc. still won't launch. When SteamVRs Direct display mode is enabled lazer won't launch. Disabling direct display mode Lets lazer Launch. Normal Osu! can launch regardless of SteamVRs Direct Display mode.

**Reply by peppy** [reply]
Relevant portion of logs:

```csharp

2022-09-26 06:38:48 [verbose]: SDL error log [debug]: Couldn't find display mode match
2022-09-26 06:38:48 [error]: An unhandled error has occurred.
2022-09-26 06:38:48 [error]: System.InvalidOperationException: couldn't retrieve valid display mode
2022-09-26 06:38:48 [error]: at osu.Framework.Platform.SDL2DesktopWindow.getClosestDisplayMode(Size size, Int32 refreshRate, Int32 displayIndex)
2022-09-26 06:38:48 [error]: at osu.Framework.Platform.SDL2DesktopWindow.updateWindowStateAndSize()
2022-09-26 06:38:48 [error]: at osu.Framework.Platform.SDL2DesktopWindow.updateWindowSpecifics()
2022-09-26 06:38:48 [error]: at osu.Framework.Platform.SDL2DesktopWindow.Run()
2022-09-26 06:38:48 [error]: at osu.Framework.Platform.GameHost.Run(Game game)
```

**Reply by peppy** [reply]
I think this will be covered by https://github.com/ppy/osu/issues/19877#issuecomment-1221438234 / https://github.com/ppy/osu-framework/pull/5371.

**Reply by M4rkFlor** [reply]
Not sure what changed but, the game launches in windowed mode. switching to fullscreen in the settings causes the game to scale to full screen but the window is still the same size as in window mode so I can only see a cropped view. pressing alt+enter forces the window to resize to fullscreen but then the game starts to lag horribly like 60hz but the fps counter is at 1000.

img for setting to fullscreen through menu.
![image](https://user-images.githubusercontent.com/38170532/194504841-c7194948-3ea1-4d6d-b6db-f02c3d53e9e5.png)
 The exit button and the top menu items are out of the viewport.

It looks like it can get a valid display mode but now its complaining about this:
```
2022-10-07 07:43:25 [verbose]: SDL error log [debug]: ChangeDisplaySettingsEx() failed: DISP_CHANGE_BADPARAM
```
also switching to borderless/windowed and back to fullscreen shows the error in the logs more.

logs:
[legacy-ipc.log](https://github.com/ppy/osu/files/9731813/legacy-ipc.log)
[performance.log](https://github.com/ppy/osu/files/9731814/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9731815/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9731816/updater.log)
[database.log](https://github.com/ppy/osu/files/9731817/database.log)

**Comment by Susko3** [comment, marked as answer]
There were recent [efforts](https://github.com/ppy/osu-framework/pull/5469) to try to fix this (and similar) crashes. Could you try again and report back?
osu! probably won't work prefect with SteamVR, but it should crash less.

**Reply by M4rkFlor** [reply]
game wont launch this error happens.
[runtime.log](https://github.com/ppy/osu/files/10095327/runtime.log)
Osu stable launches with no issue.

**Reply by Susko3** [reply]
This means that SDL is not recognizing any displays – it thinks there are 0 displays. I'm not sure how SDL displays are supposed to work with SteamVR. I would hope that Valve tested this, they seems to be involved with SDL a lot.

I don't know if this is a problem with your specific setup, or with SDL.

**Reply by Susko3** [reply]
I've added some additional logging in https://github.com/ppy/osu-framework/pull/5566. Please update and post new logs of the crash.

**Reply by M4rkFlor** [reply]
I updated, error looks the same to me.
[input.log](https://github.com/ppy/osu/files/10234011/input.log)
[runtime.log](https://github.com/ppy/osu/files/10234012/runtime.log)
[network.log](https://github.com/ppy/osu/files/10234013/network.log)
[performance.log](https://github.com/ppy/osu/files/10234014/performance.log)
[updater.log](https://github.com/ppy/osu/files/10234015/updater.log)
[database.log](https://github.com/ppy/osu/files/10234017/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/10234018/legacy-ipc.log)

I found out if I unplug everything including main monitor, then only plug in my main monitor the game launches but has a slightly different log. "SDL error log [debug]: Invalid window"

[runtime.log](https://github.com/ppy/osu/files/10234042/runtime.log)

**Reply by Susko3** [reply]
The invalid window bug should be fixed now, and with https://github.com/ppy/osu-framework/pull/5602 there should be more useful logs. Please give it another try and report back!

**Reply by M4rkFlor** [reply]
Still no launch, but new logs
[runtime.log](https://github.com/ppy/osu/files/10888013/runtime.log)

**Reply by Susko3** [reply]
Relevant line from the log:

```cs
2023-03-04 09:54:54 [verbose]: Failed to get display modes for display at index (0) (1920x1080). SDL Error:  (0)
```

So it can properly get the resolution of the display, but `SDL_GetNumDisplayModes` returned zero modes. And there is no reported error. SteamVR probably creates a virtual display that can't change modes, but it doesn't have any modes -- which is a bit strange.

I think the way to fix this is to consider zero (`0`) display modes a valid display and handle the lack of display modes elsewhere (i.e. by falling back to borderless fullscreen).

**Reply by Susko3** [reply]
I've tried to fix this (again) in https://github.com/ppy/osu-framework/pull/5665. Please give it another try and send back the logs (even if it works!).

**Reply by M4rkFlor** [reply]
[runtime.log](https://github.com/ppy/osu/files/11014914/runtime.log)
it automatically opened as a small window instead of fullscreen. I changed it to fullscreen and it actually went to full screen without scaling issues like the last time. however it says its not in exclusive fullscreen mode.

**Reply by Susko3** [reply]
There are still some errors in the logs, but since this is an unusual setup and the game launches now, I'm going to mark this as solved. If you run into any glaring issues again, feel free to open another discussion (and mention this one!).

```cs
2023-03-20 06:31:18 [verbose]: SDL error log [debug]: ChangeDisplaySettingsEx() failed: DISP_CHANGE_BADPARAM
// ...
2023-03-20 06:31:23 [verbose]: Exclusive fullscreen capability: Incapable (QUNS_ACCEPTS_NOTIFICATIONS)
```
